File: packages/java/src/test/java/ai/webpipe/sdk/WebpipeClientTest.java
package ai.webpipe.sdk;

import static org.junit.jupiter.api.Assertions.*;

import ai.webpipe.sdk.exceptions.AuthenticationException;
import ai.webpipe.sdk.exceptions.CrawlException;
import ai.webpipe.sdk.exceptions.NotFoundException;
import ai.webpipe.sdk.exceptions.PollTimeoutException;
import ai.webpipe.sdk.model.CrawlJob;
import ai.webpipe.sdk.model.Document;
import ai.webpipe.sdk.model.MapLink;
import ai.webpipe.sdk.options.CrawlOptions;
import ai.webpipe.sdk.options.MapOptions;
import ai.webpipe.sdk.options.ScrapeOptions;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

class WebpipeClientTest {

    private WebpipeClient client(MockServer server) {
        return WebpipeClient.builder()
                .apiKey("wc-test-key")
                .baseUrl(server.url())
                .maxRetries(0)
                .build();
    }

    @Test
    void scrapeSuccess() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/scrape", 200, """
                    {"success":true,"data":{"markdown":"# Example",
                      "metadata":{"title":"Example Domain","sourceURL":"https://example.com","statusCode":200}}}""");
            Document doc = client(server).scrape("https://example.com",
                    new ScrapeOptions().formats(List.of("markdown", "metadata")));

            assertEquals("# Example", doc.markdown());
            assertEquals("Example Domain", doc.metadata().get("title"));
            assertEquals(200, doc.metadata().get("statusCode"));

            MockServer.Request req = server.lastRequest();
            assertEquals("Bearer wc-test-key", req.headers().get("Authorization"));
            assertTrue(req.headers().get("User-agent").startsWith("webpipe-sdk-java/"));
            assertEquals("https://example.com", req.body().get("url").asText());
            assertEquals("markdown", req.body().get("formats").get(0).asText());
            assertFalse(req.body().has("use_proxy"), "unset options must be omitted");
        }
    }

    @Test
    void scrapeSnakeCaseMapping() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/scrape", 200, "{\"success\":true,\"data\":{\"json\":{}}}");
            client(server).scrape("https://example.com",
                    new ScrapeOptions()
                            .formats(List.of("json"))
                            .jsonSchema(Map.of("type", "object",
                                    "properties", Map.of("myField", Map.of("type", "string"))))
                            .useProxy(true));

            var body = server.lastRequest().body();
            assertTrue(body.get("use_proxy").asBoolean());
            assertTrue(body.get("json_schema").get("properties").has("myField"),
                    "user keys inside jsonSchema must pass through untouched");
        }
    }

    @Test
    void scrape401() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/scrape", 401, "{\"success\":false,\"error\":\"invalid key\"}");
            AuthenticationException e = assertThrows(AuthenticationException.class,
                    () -> client(server).scrape("https://example.com"));
            assertEquals(401, e.getStatusCode());
            assertTrue(e.getMessage().contains("invalid key"));
        }
    }

    @Test
    void retriesOn429() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/scrape", 429, "{\"success\":false,\"error\":\"slow down\"}");
            server.enqueue("POST", "/api/v1/scrape", 200, "{\"success\":true,\"data\":{\"markdown\":\"ok\"}}");
            WebpipeClient c = WebpipeClient.builder()
                    .apiKey("wc-test-key").baseUrl(server.url()).maxRetries(1).build();
            assertEquals("ok", c.scrape("https://example.com").markdown());
            assertEquals(2, server.requests.size());
        }
    }

    @Test
    void mapReturnsLinks() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/map", 200, """
                    {"success":true,"links":[{"url":"https://example.com/about","title":"About"},
                     {"url":"https://example.com/blog"}]}""");
            List<MapLink> links = client(server).map("https://example.com",
                    new MapOptions().limit(100).sameDomain(false));
            assertEquals(2, links.size());
            assertEquals("https://example.com/about", links.get(0).url());
            assertFalse(server.lastRequest().body().get("same_domain").asBoolean());
        }
    }

    @Test
    void crawlPollsUntilCompleted() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/crawl", 200, "{\"success\":true,\"id\":\"job-1\"}");
            server.enqueue("GET", "/api/v1/crawl/job-1", 200,
                    "{\"success\":true,\"status\":\"scraping\",\"total\":2,\"completed\":1}");
            server.enqueue("GET", "/api/v1/crawl/job-1", 200, """
                    {"success":true,"status":"completed","total":2,"completed":2,
                     "data":[{"markdown":"# A"},{"markdown":"# B"}],"errors":[],
                     "expires_at":"2026-07-28T00:00:00+00:00"}""");
            CrawlJob job = client(server).crawl("https://x",
                    new CrawlOptions().limit(2), 0.01, 0);
            assertEquals("completed", job.status());
            assertEquals(2, job.data().size());
            assertEquals("# B", job.data().get(1).markdown());
            assertEquals("2026-07-28T00:00:00+00:00", job.expiresAt());
        }
    }

    @Test
    void crawlFailed() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/crawl", 200, "{\"success\":true,\"id\":\"job-2\"}");
            server.enqueue("GET", "/api/v1/crawl/job-2", 200,
                    "{\"success\":true,\"status\":\"failed\",\"errors\":[\"boom\"]}");
            assertThrows(CrawlException.class,
                    () -> client(server).crawl("https://x", new CrawlOptions(), 0.01, 0));
        }
    }

    @Test
    void crawlTimeout() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/crawl", 200, "{\"success\":true,\"id\":\"job-3\"}");
            for (int i = 0; i < 20; i++) {
                server.enqueue("GET", "/api/v1/crawl/job-3", 200,
                        "{\"success\":true,\"status\":\"scraping\"}");
            }
            assertThrows(PollTimeoutException.class,
                    () -> client(server).crawl("https://x", new CrawlOptions(), 0.01, 0.05));
        }
    }

    @Test
    void crawlStatus404() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("GET", "/api/v1/crawl/gone", 404, "{\"success\":false,\"error\":\"gone\"}");
            assertThrows(NotFoundException.class, () -> client(server).getCrawlStatus("gone"));
        }
    }

    // -------------------------------------------------------------- async

    @Test
    void scrapeAsyncSuccess() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/scrape", 200,
                    "{\"success\":true,\"data\":{\"markdown\":\"# Async\"}}");
            Document doc = client(server).scrapeAsync("https://example.com",
                    new ScrapeOptions().formats(List.of("markdown"))).join();
            assertEquals("# Async", doc.markdown());
        }
    }

    @Test
    void scrapeAsync401() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/scrape", 401, "{\"success\":false,\"error\":\"invalid key\"}");
            var future = client(server).scrapeAsync("https://example.com", null);
            var thrown = assertThrows(java.util.concurrent.CompletionException.class, future::join);
            assertTrue(thrown.getCause() instanceof AuthenticationException);
            assertEquals("invalid key",
                    ((AuthenticationException) thrown.getCause()).getMessage().contains("invalid key")
                            ? "invalid key" : "");
        }
    }

    @Test
    void crawlAsyncPollsUntilCompleted() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/crawl", 200, "{\"success\":true,\"id\":\"job-a1\"}");
            server.enqueue("GET", "/api/v1/crawl/job-a1", 200,
                    "{\"success\":true,\"status\":\"scraping\",\"total\":2,\"completed\":1}");
            server.enqueue("GET", "/api/v1/crawl/job-a1", 200,
                    "{\"success\":true,\"status\":\"completed\",\"total\":2,\"completed\":2,"
                            + "\"data\":[{\"markdown\":\"# A\"}],\"errors\":[]}");
            CrawlJob job = client(server).crawlAsync("https://x",
                    new CrawlOptions().limit(2), 0.01, 0).join();
            assertEquals("completed", job.status());
            assertEquals(1, job.data().size());
        }
    }
}

File: packages/php/tests/WebpipeClientTest.php
<?php

declare(strict_types=1);

namespace Webpipe\Tests;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Middleware;
use GuzzleHttp\Psr7\Response;
use PHPUnit\Framework\TestCase;
use Webpipe\CrawlJob;
use Webpipe\Exception\AuthenticationException;
use Webpipe\Exception\CrawlException;
use Webpipe\Exception\NotFoundException;
use Webpipe\Exception\PollTimeoutException;
use Webpipe\Internal\HttpClient;
use Webpipe\WebpipeClient;

final class WebpipeClientTest extends TestCase
{
    /** @var list<array{request: mixed, response: mixed}> */
    private array $history = [];

    private function mockClient(array $responses, int $maxRetries = 0): WebpipeClient
    {
        $mock = new MockHandler($responses);
        $stack = HandlerStack::create($mock);
        $stack->push(Middleware::history($this->history));
        $guzzle = new GuzzleClient(['handler' => $stack, 'http_errors' => false]);
        $http = HttpClient::create(
            'https://api.web2json.ai',
            'wc-test-key',
            60.0,
            $maxRetries,
            'webpipe-sdk-php/' . WebpipeClient::VERSION,
            $guzzle,
        );
        return new WebpipeClient(apiKey: 'wc-test-key', httpClient: $http);
    }

    private static function json(array $body, int $status = 200): Response
    {
        return new Response($status, ['Content-Type' => 'application/json'], (string) json_encode($body));
    }

    public function testMissingApiKeyThrows(): void
    {
        putenv('WEBPIPE_API_KEY');
        $this->expectException(\InvalidArgumentException::class);
        $this->expectExceptionMessageMatches('/WEBPIPE_API_KEY/');
        new WebpipeClient();
    }

    public function testScrapeSuccess(): void
    {
        $client = $this->mockClient([
            self::json([
                'success' => true,
                'data' => [
                    'markdown' => '# Example',
                    'metadata' => ['title' => 'Example Domain', 'sourceURL' => 'https://example.com', 'statusCode' => 200],
                ],
            ]),
        ]);

        $doc = $client->scrape('https://example.com', formats: ['markdown', 'metadata']);

        self::assertSame('# Example', $doc->markdown);
        self::assertSame('Example Domain', $doc->metadata['title']);
        self::assertSame(200, $doc->metadata['statusCode']);

        $request = $this->history[0]['request'];
        self::assertSame('Bearer wc-test-key', $request->getHeaderLine('Authorization'));
        self::assertStringStartsWith('webpipe-sdk-php/', $request->getHeaderLine('User-Agent'));
        $body = json_decode((string) $request->getBody(), true);
        self::assertSame(['url' => 'https://example.com', 'formats' => ['markdown', 'metadata']], $body);
    }

    public function testScrape401(): void
    {
        $client = $this->mockClient([
            self::json(['success' => false, 'error' => 'invalid key'], 401),
        ]);
        try {
            $client->scrape('https://example.com');
            self::fail('expected AuthenticationException');
        } catch (AuthenticationException $e) {
            self::assertSame(401, $e->statusCode);
            self::assertStringContainsString('invalid key', $e->getMessage());
        }
    }

    public function testScrapeRetriesOn429(): void
    {
        $client = $this->mockClient([
            self::json(['success' => false, 'error' => 'slow down'], 429),
            self::json(['success' => true, 'data' => ['markdown' => 'ok']]),
        ], maxRetries: 1);

        $doc = $client->scrape('https://example.com');
        self::assertSame('ok', $doc->markdown);
        self::assertCount(2, $this->history);
    }

    public function testMapReturnsLinks(): void
    {
        $client = $this->mockClient([
            self::json([
                'success' => true,
                'links' => [
                    ['url' => 'https://example.com/about', 'title' => 'About'],
                    ['url' => 'https://example.com/blog'],
                ],
            ]),
        ]);
        $links = $client->map('https://example.com', limit: 100, sameDomain: false);
        self::assertCount(2, $links);
        self::assertSame('https://example.com/about', $links[0]->url);

        $body = json_decode((string) $this->history[0]['request']->getBody(), true);
        self::assertFalse($body['same_domain']);
        self::assertSame(100, $body['limit']);
    }

    public function testCrawlPollsUntilCompleted(): void
    {
        $client = $this->mockClient([
            self::json(['success' => true, 'id' => 'job-1']),
            self::json(['success' => true, 'status' => 'scraping', 'total' => 2, 'completed' => 1]),
            self::json([
                'success' => true, 'status' => 'completed', 'total' => 2, 'completed' => 2,
                'data' => [['markdown' => '# A'], ['markdown' => '# B']],
                'errors' => [],
                'expires_at' => '2026-07-28T00:00:00+00:00',
            ]),
        ]);

        $job = $client->crawl('https://x', limit: 2, pollInterval: 0.01);

        self::assertSame(CrawlJob::STATUS_COMPLETED, $job->status);
        self::assertCount(2, $job->data);
        self::assertSame('# B', $job->data[1]->markdown);
        self::assertSame('2026-07-28T00:00:00+00:00', $job->expiresAt);
    }

    public function testCrawlFailed(): void
    {
        $client = $this->mockClient([
            self::json(['success' => true, 'id' => 'job-2']),
            self::json(['success' => true, 'status' => 'failed', 'errors' => ['boom']]),
        ]);
        $this->expectException(CrawlException::class);
        $client->crawl('https://x', pollInterval: 0.01);
    }

    public function testCrawlTimeout(): void
    {
        // first response is the submit; subsequent GETs always "scraping"
        $mock = new MockHandler(array_merge(
            [self::json(['success' => true, 'id' => 'job-3'])],
            array_fill(0, 50, self::json(['success' => true, 'status' => 'scraping'])),
        ));
        $stack = HandlerStack::create($mock);
        $guzzle = new GuzzleClient(['handler' => $stack, 'http_errors' => false]);
        $http = HttpClient::create(
            'https://api.web2json.ai',
            'wc-test-key',
            60.0,
            0,
            'webpipe-sdk-php/' . WebpipeClient::VERSION,
            $guzzle,
        );
        $client = new WebpipeClient(apiKey: 'wc-test-key', httpClient: $http);

        $this->expectException(PollTimeoutException::class);
        $client->crawl('https://x', pollInterval: 0.01, timeout: 0.05);
    }

    public function testGetCrawlStatus404(): void
    {
        $client = $this->mockClient([
            self::json(['success' => false, 'error' => 'gone'], 404),
        ]);
        $this->expectException(NotFoundException::class);
        $client->getCrawlStatus('gone');
    }

    public function testCrawlProgressCallback(): void
    {
        $client = $this->mockClient([
            self::json(['success' => true, 'id' => 'job-9']),
            self::json(['success' => true, 'status' => 'scraping', 'total' => 3, 'completed' => 1]),
            self::json(['success' => true, 'status' => 'scraping', 'total' => 3, 'completed' => 2]),
            self::json(['success' => true, 'status' => 'completed', 'total' => 3, 'completed' => 3, 'data' => [], 'errors' => []]),
        ]);

        $progress = [];
        $job = $client->crawl(
            'https://x',
            limit: 3,
            pollInterval: 0.01,
            onProgress: function (CrawlJob $status) use (&$progress): void {
                $progress[] = "{$status->completed}/{$status->total}";
            },
        );

        self::assertSame(CrawlJob::STATUS_COMPLETED, $job->status);
        self::assertSame(['1/3', '2/3'], $progress);
    }
}
